cmd/lambda/accounts/create.go

```diff
@@ -2,251 +2,31 @@ package main
-	isValid := true
-	var validationErrors []error
-	if req.ID == "" {
-		isValid = false
-		validationErrors = append(validationErrors, errors.New("missing required field \"id\""))
-	}
-	if req.AdminRoleArn == "" {
-		isValid = false
-		validationErrors = append(validationErrors, errors.New("missing required field \"adminRoleArn\""))
-	}
-
-	if !isValid {
-		errMsgs := []string{}
-		for _, verr := range validationErrors {
-			errMsgs = append(errMsgs, verr.Error())
-		}
-		msg := strings.Join(errMsgs, "; ")
-		return false, &msg
-	}
-
-	return true, nil
-}
-
-func createPrincipalRole(childAccount db.Account, masterAccountID string) (*rolemanager.CreateRoleWithPolicyOutput, string, error) {
-	// Create an assume role policy,
-	// to let principals from the same account assume the role.
-	//
-	// Consumers of open source DCE may modify and customize
-	// this as need (eg. to integrate with SSO/SAML)
-	// by responding to the "account-created" SNS topic
-	assumeRolePolicy := strings.TrimSpace(fmt.Sprintf(`
-		{
-			"Version": "2012-10-17",
-			"Statement": [
-				{
-					"Effect": "Allow",
-					"Principal": {
-						"AWS": "arn:aws:iam::%s:root"
-					},
-					"Action": "sts:AssumeRole",
-					"Condition": {}
-				}
-			]
-		}
-	`, masterAccountID))
-
-	// Render the default policy for the principal
-
-	policy, policyHash, err := StorageSvc.GetTemplateObject(artifactsBucket, principalPolicyS3Key,
-		principalPolicyInput{
-			PrincipalPolicyArn:   fmt.Sprintf("arn:aws:iam::%s:policy/%s", childAccount.ID, policyName),
-			PrincipalRoleArn:     fmt.Sprintf("arn:aws:iam::%s:role/%s", childAccount.ID, principalRoleName),
-			PrincipalIAMDenyTags: principalIAMDenyTags,
-			AdminRoleArn:         childAccount.AdminRoleArn,
-			Regions:              allowedRegions,
-		})
-	if err != nil {
-		return nil, "", err
-	}
-
-	// Assume role into the new account
-	accountSession, err := TokenSvc.NewSession(AWSSession, childAccount.AdminRoleArn)
-	if err != nil {
-		return nil, "", err
-	}
-	iamClient := iam.New(accountSession)
-
-	// Create the Role + Policy
-	RoleManager.SetIAMClient(iamClient)
-	createRoleOutput, err := RoleManager.CreateRoleWithPolicy(&rolemanager.CreateRoleWithPolicyInput{
-		RoleName:                 principalRoleName,
-		RoleDescription:          "Role to be assumed by principal users of DCE",
-		AssumeRolePolicyDocument: assumeRolePolicy,
-		MaxSessionDuration:       principalMaxSessionDuration,
-		PolicyName:               policyName,
-		PolicyDocument:           policy,
-		PolicyDescription:        "Policy for principal users of DCE",
-		Tags: append(tags,
-			&iam.Tag{Key: aws.String("Name"), Value: aws.String("DCEPrincipal")},
-		),
-		IgnoreAlreadyExistsErrors: true,
-	})
-	return createRoleOutput, policyHash, err
-}
-
-type principalPolicyInput struct {
-	PrincipalPolicyArn   string
-	PrincipalRoleArn     string
-	PrincipalIAMDenyTags []string
-	AdminRoleArn         string
-	Regions              []string
```

Comment:
Does it keep you up at night, @kddejong, the number of lines of code you've killed with your bare hands? 😜

Reply:
Seriously though, this is really cool, how clean this controller becomes